server/utils/notifier.py

```diff
@@ -0,0 +1,177 @@
+so on.  The purpose of these classes is to reduce code for sending messages
+to specific subsets of the game.  (e.g. - send only to staff in a given room)
+
+USAGE (example code):
+
+# This code will send "Hello, world!" then 'msg' to all player GMs
+# and staff GMs in the given room.
+gm_notifier = RoomNotifier(caller, room=caller.location, to_gm=True, to_staff=True)
+gm_notifier.notify("Hello, world!")
+gm_notifier.notify(msg)
+"""
+from typing import List
+
+from typeclasses.rooms import ArxRoom
+
+
+class NotifyError(Exception):
+    pass
+
+
+class Notifier:
+    """
+    Base class for sending notifications to the game.
+    This class is meant to be derived from and its ('private') code
+    utilized in derived classes (to lower code duplication).
+    """
+
+    def __init__(
+        self,
+        caller,
+        **to_flags,
+    ):
+        self.caller = caller
+        self.to_flags = to_flags
+
+        self.receiver_set = set()
+
+    def notify(self, msg: str, options: dict = None):
+        """Notifies each receiver of msg with the given options, if any."""
+        for rcvr in self.receiver_set:
+            rcvr.msg(msg, options)
+
+    @property
+    def receivers(self) -> set:
+        return self.receiver_set
+
+    @property
+    def receiver_names(self) -> List[str]:
+        return [str(player) for player in self.receiver_set]
+
+    def generate(self):
+        pass
+
+    def _filter_players(self) -> set:
+        """Returns all non-gm, non-staff players in receiver_set."""
+        player_set = {
+            char for char in self.receiver_set if not char.check_staff_or_gm()
+        }
+        return player_set
+
+    def _filter_gms(self) -> set:
+        """Returns all player GMs in receiver_set."""
+        gm_set = {
+            char
+            for char in self.receiver_set
+            if char.check_staff_or_gm() and not char.check_permstring("builder")
+        }
+        return gm_set
+
+    def _filter_staff(self) -> set:
+        """Returns all staff in receiver_set."""
+        staff_set = {
+            char for char in self.receiver_set if char.check_permstring("builder")
+        }
+        return staff_set
+
+    def _filter_receivers(self):
+        """Returns all receivers designated by the given receiver flags."""
+        player_set = set()
+        gm_set = set()
+        staff_set = set()
+
+        if self.to_flags.get("to_player", False):
+            player_set = self._filter_players()
+
+        if self.to_flags.get("to_gm", False):
+            gm_set = self._filter_gms()
+
+        if self.to_flags.get("to_staff", False):
+            staff_set = self._filter_staff()
+
+        self.receiver_set = player_set | gm_set | staff_set
+
+
+class RoomNotifier(Notifier):
+    """
+    Notifier for sending to everyone in a room, filtered by
+    the to_ flags.
+    """
+
+    def __init__(
+        self,
+        caller,
+        room: ArxRoom,
```

Comment:
I'd rather not have an import of a typeclass in a utility file because it'd be really easy to have circular imports occur. You can do a lazy reference of "ArxRoom" most likely, which a lot of linters will fail to resolve, or just skip it entirely. I just really hate any imports of other project modules inside a utility module if you can avoid it